### PR TITLE
Fix duplication and whitesource

### PR DESCRIPTION
**Pull request by MarianG**
**Note: DefectDojo is now on Python3 and Django 2.2.1 Please submit your pull requests to the 'dev' branch as the 'legacy-python2.7' branch is only for bug fixes. Any new features submitted to the legacy branch will be ignored and closed.**

The change is to fix a problem, where we found circular dependencies in duplicates of findings. We adapted the set_duplicate method to make sure circular dependencies will not occur anymore. Additionally we included a method to find and fix loops in the duplicates.
Unit-Tests are provided along with the changes

Furthermore the whitesource-change with the title resulted in the problem that new whitesource findings have different titles and thus different hashes, which makes the duplication not applicable.
A fix for this is provided in the utils.

When submitting a pull request, please make sure you have completed the following checklist:

- [x ] Your code is flake8 compliant 
- [x] Your code is python 3.5 compliant
- [ ] If this is a new feature and not a bug fix, you've included the proper documentation in the ReadTheDocs documentation folder. https://github.com/DefectDojo/Documentation/tree/master/docs or provide feature documentation in the PR.
- [ ] Model changes must include the necessary migrations in the dojo/dd_migrations folder.
- [ ] Add applicable tests to the unit tests.

### CODE REVIEW

**Comment by MarianG** [comment]
The change is to fix a problem, where we found circular dependencies in duplicates of findings. We adapted the set_duplicate method to make sure circular dependencies will not occur anymore. Additionally we included a method to find and fix loops in the duplicates. 
Unit-Tests are provided along with the changes

Furthermore the whitesource-change with the title resulted in the problem that new whitesource findings have different titles and thus different hashes, which makes the duplication not applicable.
A fix for this is provided in the utils.

File: dojo/models.py [hunk 1]
```diff
@@ -1226,7 +1226,7 @@ def __str__(self):
 
 
 class Finding(models.Model):
-    title = models.CharField(max_length=511)
+    title = models.TextField(max_length=511)
```
Comment (valentijnscholten) [inline dojo/models.py]: I think the type was changed to CharField for a reason (index?) so unless there is a good reason, it should stay CharField?

File: dojo/models.py [hunk 2]
```diff
@@ -1670,6 +1674,11 @@ def save(self, dedupe_option=True, false_history=False, rules_option=True, issue
         calculate_grade(self.test.engagement.product)
 
     def delete(self, *args, **kwargs):
+        if self.id:
+            logger.info("Debug-Delete: Deleting Finding with id: %d", self.id)
```
Comment (valentijnscholten) [inline dojo/models.py]: suggest to log to debug or even lower

File: dojo/utils.py [hunk 3]
```diff
@@ -544,116 +641,60 @@ def add_breadcrumb(parent=None,
...
+                            finding.date.weekday()]]
+            except:
+                if new_date < finding.date <= end_date:
+                    # [0,0,(20*.02)]
+                    # [week, day, weight]
+                    days[day_offset[finding.date.weekday()]] += 1
+                    if days[day_offset[finding.date.weekday()]] > highest_count:
+                        highest_count = days[day_offset[
+                            finding.date.weekday()]]
+                pass
```
Comment (valentijnscholten) [inline dojo/utils.py]: can you rebase your branch, looks like you are overwriting changes with older stuff

**Comment by valentijnscholten** [comment]
Could you explain when this circular referencing with duplicates could occur? It sounds to me more like maybe some corners case in the deduplication we need to fix to prevent this.
Also (to be sure) could you rebase your branch/changes? There are some changes made in dev that are being overwritten back to their old version in your PR.

**Comment by MarianG** [comment]
Hi
we found some additional bugs and typos for the whitesource thing.
Can I close this PR and recreate it once we are ready with a better quality?

**Pull request closed without merging**